nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java

```diff
@@ -114,20 +123,48 @@ protected void doTrashDocument(DocumentModel doc) {
         }
         session.setDocumentSystemProp(docRef, SYSPROP_IS_TRASHED, Boolean.FALSE);
         notifyEvent(session, DOCUMENT_UNTRASHED, docForEvent);
+        if (processChildren && session.hasChildren(doc.getRef())) {
+            setTrashPropertyToChildren(doc, Boolean.FALSE);
+        }
 
         Set<DocumentRef> docRefs = new HashSet<>();
         docRefs.add(docRef);
 
         // now untrash the parent if needed
         if (parentRef != null && session.isTrashed(parentRef)) {
             DocumentModel parent = session.getDocument(parentRef);
-            parent.putContextData(SKIP_CHILDREN_PROCESSING_KEY, Boolean.TRUE);
-            Set<DocumentRef> ancestorDocRefs = doUntrashDocument(parent);
+            Set<DocumentRef> ancestorDocRefs = doUntrashDocument(parent, false);
             docRefs.addAll(ancestorDocRefs);
         }
         return docRefs;
     }
 
+    protected void setTrashPropertyToChildren(DocumentModel model, Boolean value) {
+        String eventId = value ? DOCUMENT_TRASHED : DOCUMENT_UNTRASHED;
+        String currentTrashValue = value ? "0" : "1";
+        CoreSession session = model.getCoreSession();
+        BulkService service = Framework.getService(BulkService.class);
+        String nxql = String.format(
+                "SELECT * from Document where ecm:isProxy = 0 AND ecm:isTrashed = %s AND ecm:ancestorId='%s'",
+                currentTrashValue, model.getId());
```

Comment:
Probably not needed right now, but can be cool if we do it on the list of documents of trash/untrash methods if it's easily possible.